README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

[![R build
status](https://github.com/obrl-soil/em38/workflows/R-CMD-check/badge.svg)](https://github.com/obrl-soil/em38/actions)
[![R build
status](https://github.com/obrl-soil/em38/workflows/test-coverage/badge.svg)](https://github.com/obrl-soil/em38/actions)
[![R build
status](https://github.com/obrl-soil/em38/workflows/pkgdown/badge.svg)](https://github.com/obrl-soil/em38/actions)
[![Coverage
status](https://codecov.io/gh/obrl-soil/em38/branch/master/graph/badge.svg)](https://codecov.io/github/obrl-soil/em38?branch=master)
[![License:
MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

# em38

em38 translates \*.n38 binary files generated by the [Geonics EM38-MKII
ground conductivity meter](http://geonics.com/html/em38.html), commonly
used in soil research and precision agriculture. EM38-MKII devices are
supported by an official translation app called DAT38MK2, but this lacks
a command-line interface and is not easy to use in a reproducible
workflow.

## Installation

Install from github with

``` r
library(devtools)
install_github("obrl-soil/em38")
```

This package is in early development phase. Improving its reliability
requires access to a wide variety of test datasets, so if you have an
\*.n38 file that fails to decode correctly, please consider sending it
to me. I have been able to test decoding on around 50 sample files from
two different EM38-MKII devices, but could always use more. The device
and its logger have a large number of possible setting combinations, and
I have not seen all of them. Additionally, a third-party GPS needs to be
attached to the device and its data-logger, and GPS output data are
notoriously variable by brand and model.

## Usage

em38 is accompanied by a demo dataset gathered during a training
exercise:

``` r
library(em38)

# all-in-one wrapper function:
demo_survey  <- em38_from_file(path = system.file("extdata", "em38_demo.n38", 
                                                  package = "em38"),
                               hdop_filter = 3)


# Plot spatialised output of survey line 1 (calibrated conductivity for coil
# separation 0.5m)
sl1 <- demo_survey$survey_lines[[1]]
plot(sl1[sl1$mode == 'Vertical', 'cond_05'], pch = 20, 
     main = 'Conductivity, Vertical Dipole Mode, Coil Separation 0.5m')
```

![](man/figures/README-example-1.png)<!-- -->

Its surprisingly difficult to walk in a straight line across a paddock
:no_mouth:

If you want to look at the intermediate data more closely,

``` r
# import binary file as raw() type matrix
n38_mat     <- n38_import(system.file("extdata", "em38_demo.n38", package = "em38"))
# break matrix into sections according to file spec 
n38_chunks  <- n38_chunk(n38_mat)
# decode matrix chunks into useable data
n38_decoded <- n38_decode(n38_chunks)
```

You can also create the equivalent of an \*.M38 text file, for
comparison.

``` r
m38_example <- n38_to_m38(n38_decoded)
# write to file as e.g.
#  write(m38_example, paste0('m38_from_R_', Sys.Date(), '.m38'))
```

------------------------------------------------------------------------


NEWS.md
# Version 0.0.6

  * some under-the-hood corrections to work with recent R changes
  * some improvements in handling data from certain GNSS receivers, many thanks to the person who supplied a test dataset :pray:
  * updated vignette, various methods of spatially interpolating the raw data are now demonstrated

# Version 0.0.5

  * package now reports orthometric height of each spatialised reading, calculated from GPS data, per user request (thanks dukarvat!). Note that elevation data may still need further correction to a local vertical datum before use.

# Version 0.0.4

  * package can now handle GPS messages from GLONASS, GALILEO, and multi-GNSS
  systems

# Version 0.0.3

  * Update for compatibility with newer versions of `units`

# Version 0.0.2

  * Bugfix for GH-1 comment processing, many thanks to sanjaykmenon for 
    providing test data  

# Version 0.0.1

  * Major refactor - process survey lines separately, cope with missing/bad GPS
    data, better paired data handling
  * Added `em38_surveyline()` which will process data from an individual survey 
    line output by `n38_decode()`.
  * Survey line decodes can now be `sf` or `data.frame` and include both 
    timestamps and sequential ID.
  * renamed `em38_spatial()` to `em38_decode()`, the function now outputs a list 
    containing file header data and survey lines.
  * renamed `n38_to_points()` to `em38_from_file()`, the output is identical to 
    `em38_decode()`  
  * Survey lines are no longer filtered by dipole mode
  * Added filters for GPS time delay and signal quality.
  * Various small bugfixes

# Version 0.0.0.9003

  * `n38_decode()` better living through proper vectorisation - speed 
    improvement.
  * More unit tests.
  * Package-level metadata.
  * Travis-CI integration

# Version 0.0.0.9002

  * New `em38_pair()` - combines data from manual-mode surveys where 
    horizontal and vertical readings have been taken at each station.
  * Checksums on NMEA-0183 sentences are evaluated for $GPGGA type, failures 
    are dropped.
  * Who's got two thumbs and forgot about `@keywords Internal`? \*gestures at 
    self\*. The help menu is less cluttered now.

# Version 0.0.0.9001

  * New `n38_to_points()` wrapper function - on-disk file to spatial points 
    in one hit.
  * `n38_import()` variable name fix; `x` is now `path`.
  * `em38_spatialise()` now constrains `out_mode` values and fails gracefully 
    when allowable values aren't supplied.
  * Rebuilt demo data to match demo extdata.

# Version 0.0.0.9000

  * Basic functions: import N38 files, convert them to a nested list of raw() 
    matrices, decode those, output an M38 file, output a spatial point dataset.
  * Minimal testing for NMEA-0183 sentence parsing (just checks on lat and 
    long extraction from string).
  


tests/testthat.R
library(testthat)
library(em38)

test_check("em38")


tests/testthat/test_spatialise.R
context('final processing for decoded data')

test_that(
  'get_loc_data works correctly',
  c(
    data('n38_demo'),
    n38_chunks  <- n38_chunk(n38_demo),
    n38_decoded <- n38_decode(n38_chunks),
    loc_1 <- n38_decoded$survey_line_1$location_data[1:7, ],
    expect_is(get_loc_data(loc_1), 'data.frame'),
    expect_equal(names(get_loc_data(loc_1)),
                 c('LATITUDE', 'LONGITUDE', 'FIX', 'HDOP', 'ELEVATION_M',
                   'CHKSUM', 'timestamp_ms')),
    expect_equal(dim(get_loc_data(loc_1))[1], 1),
    expect_equal(dim(get_loc_data(loc_1))[2], 7),
    expect_is(get_loc_data(loc_1)$LATITUDE, 'numeric'),
    expect_is(get_loc_data(loc_1)$LONGITUDE, 'numeric'),
    expect_is(get_loc_data(loc_1)$HDOP, 'numeric'),
    expect_is(get_loc_data(loc_1)$ELEVATION_M, 'units'),
    expect_is(get_loc_data(loc_1)$CHKSUM, 'logical'),
    expect_is(get_loc_data(loc_1)$timestamp_ms, 'numeric')
    )
)

test_that(
  'em38_surveyline works correctly',
  c(
    data('n38_demo'),
    n38_chunks  <- n38_chunk(n38_demo),
    n38_decoded <- n38_decode(n38_chunks),
    sl1 <- em38_surveyline(n38_decoded[[2]], 3),
    expect_is(sl1, 'sf'),
    expect_equal(dim(sl1)[1], 3164L),
    expect_equal(dim(sl1)[2], 13L),
    expect_equal(sf::st_crs(sl1), sf::st_crs(4326)),
    expect_equal(sf::st_bbox(sl1)[[1]], 151.4341589),
    # no loc data
    sl2 <- n38_decoded[[2]],
    sl2[['location_data']] <- sl2[['location_data']][0, ],
    sl2 <- em38_surveyline(sl2, 3),
    expect_is(sl2, 'data.frame'),
    expect_equal(dim(sl2)[1], 3164L),
    expect_equal(dim(sl2)[2], 11L),
    # no reading_data
    sl3 <- n38_decoded[[2]],
    sl3[['reading_data']] <- sl3[['reading_data']][0, ],
    sl3 <- em38_surveyline(sl3, 3),
    expect_is(sl3, 'character'),
    expect_equal(sl3, "This survey line contains no readings."),
    # nuffin
    sl4 <- n38_decoded[[2]],
    sl4[['reading_data']] <- sl4[['reading_data']][0, ],
    sl4[['location_data']] <- sl4[['location_data']][0, ],
    sl4 <- em38_surveyline(sl4, 3),
    expect_is(sl4, 'character'),
    expect_equal(sl4, "This survey line contains no readings or location data."),
    # warning on no cal,
    sl5 <- n38_decoded[[2]],
    sl5[['cal_data']] <- sl5[['cal_data']][0, ],
    expect_warning(em38_surveyline(sl5, 3)),
    # test no good HDOP
    expect_equal(em38_surveyline(n38_decoded[[2]], 0.1),
                 'No readings with acceptable HDOP could be retrieved from this survey line.')
  )
)

test_that(
  'em38_decode works correctly',
  c(
    data('n38_demo'),
    n38_chunks  <- n38_chunk(n38_demo),
    n38_decoded <- n38_decode(n38_chunks),
    survey <- em38_decode(n38_decoded),
    expect_is(survey, 'list'),
    expect_length(survey, 2),
    expect_equal(names(survey), c('file_header', 'survey_lines')),
    expect_is(survey[[2]], 'list'),
    expect_equal(names(survey[[2]]), c('1'))
  )
)

test_that(
  'em38_from_file works',
  c(
    from_file <-
      em38_from_file(path = system.file("extdata", "em38_demo.N38",
                                       package = "em38"), hdop_filter = 3),
    data('n38_demo'),
    n38_chunks  <- n38_chunk(n38_demo),
    n38_decoded <- n38_decode(n38_chunks),
    survey <- em38_decode(n38_decoded),
    expect_equal(from_file, survey)
    )
)


tests/testthat/test_nmea_parsing.R
context('NMEA message parsing')

# these functions are all in GPS_data.R

# 1. nmea_check
test_that(
  'nmea checksum validator works correctly',
  c(
    msg <- "$GPGGA,015808.00,2726.53758,S,15126.05255,E,1,08,1.0,365.1,M,39.5,M,,*79",
    expect_true(nmea_check(string = msg)),
    expect_false(nmea_check(string = gsub('015', '009', msg))),
    expect_false(nmea_check(string = gsub('\\*', '', msg))),
    expect_length(nmea_check(string = msg), 1L),
    expect_type(nmea_check(string = msg), 'logical')
  )
)

# 2. process_gga
test_that(
  'process_gga works correctly',
  c(
    msg <- "$GPGGA,015808.00,2726.53758,S,15126.05255,E,1,08,1.0,365.1,M,39.5,M,,*79",
    expect_type(process_gga(msg), 'list'),
    expect_length(process_gga(msg), 11L),
    expect_is(process_gga(msg)[[1]], class = 'POSIXct'),
    expect_equal(process_gga(msg)$base_stn, NA)
  )
)

# 3. process_vtg
test_that(
  'process_vtg works correctly',
  c(
    msg <- "$GPVTG,208.02,T,,M,0.32,N,0.59,K,A*38",
    expect_type(process_vtg(msg), 'list'),
    expect_length(process_vtg(msg), 6L),
    expect_is(process_vtg(msg)$speed_kmh, 'units'),
    expect_equal(process_vtg(msg)$fix_type, 'A')
  )
)

# 4. process_rmc
test_that(
  'process_rmc works correctly',
  c(
    msg <- "$GPRMC,015808.00,A,2726.53758,S,15126.05255,E,0.32,208.02,160318,,,A*48",
    expect_type(process_rmc(msg), 'list'),
    expect_length(process_rmc(msg), 10L),
    expect_is(process_rmc(msg)$UTC_date_time, 'POSIXct'),
    expect_equal(format(process_rmc(msg)$UTC_date_time), '2018-03-16 01:58:08'),
    expect_equal(process_rmc(msg)$fix_type, 'A')
  )
)

# 5. process_gsa
test_that(
  'process_gsa works correctly',
  c(
    msg <- "$GPGSA,M,3,05,10,15,16,20,21,26,29,,,,,1.6,1.0,1.2*32",
    expect_type(process_gsa(msg), 'list'),
    expect_length(process_gsa(msg), 8L),
    expect_equal(process_gsa(msg)$VDOP, 1.2)
  )
)

# 6. process_gsv
test_that(
  'process_gsv works correctly',
  c(
    msg <- "$GPGSV,3,1,11,05,14,138,46,10,14,316,37,12,04,012,,13,24,100,*76",
    expect_type(process_gsv(msg), 'list'),
    expect_length(process_gsv(msg), 19L),
    expect_equal(process_gsv(msg)$sat_4_SNR, NA_real_)
  )
)


R/data.R
#' Data: imported N38 track file
#'
#' A dataset representing a track walked by a worker carrying an EM38-MK2
#' instrument connected to an Allegra-CX datalogger and GPS.
#'
#' @format A matrix of raw bytes, 25 columns by 37249 rows, created by calling
#'   \code{\link{n38_import}} on \code{system.file("extdata", "em38_demo.n38",
#'   package = "em38")}.
#' @source This dataset was recorded on a farm near Jondaryan, South East
#'   Queensland in March 2018 during a training exercise. The track mostly
#'   covered fallow ground but ventured into a mature sorghum crop for a few
#'   meters on the eastern edge. The logger recorded an EM reading every 0.2
#'   seconds and a GPS reading every second. The EM38 instrument was properly
#'   calibrated for quad-phase readings but may not have been for in-phase. The
#'   intent of the exercise was to measure soil water content. The soil type at
#'   this location is a Black Vertosol under the Australian Soils Classification
#'   (Isbell, 2002).
#'
"n38_demo"


R/signal_conversion.R
#' Signal to conductivity/inphase reading
#'
#' This function converts an EM38-MKII signal recieved on any of Channels 1 to 4.
#' Additional conversion is required for inphase signals, and further
#' application of a calibration factor is required to derive the final
#' conductivity or resistivity value. Do not use this function in isolation.
#' @param signal Integer.
#' @return Numeric; uncalibrated instrument reading.
#' @keywords internal
#' @examples
#' channel_1_or_3 <- em38:::get_cond(30456)
#' channel_2 <- em38:::get_cond(30456) * 0.00720475
#' channel_4 <- em38:::get_cond(30456) * 0.028819
#'
get_cond <- function(signal = NULL) {
  (signal * 5 / 1024 - 160) * 8
}

#' Signal to temperature reading
#'
#' This function calculates temperature from an EM38-MKII signal recieved on
#' Channel 5 or 6.
#' @param signal Integer.
#' @return Temperature in degrees C
#' @keywords internal
#' @note The device's effective reporting resolution is ~0.32 degrees C.
#' @examples
#' channel_5 <- em38:::get_temp(227)
#'
get_temp <- function(signal = NULL) {
  # J. Pawlowski (Geonics, Geomar), pers. comm. 2018-03-28
  signal / 3.108 - 50
}

#' G*GGA latitude
#'
#' This function retrieves latitude position information from a G*GGA string.
#' @param lat String containing latitude information. G*GGA strings are
#'   comma-separated, latitude is in the second delimited position
#' @param dir String determining whether latitude is N or S of the equator.
#'   G*GGA strings are comma-separated, NS status is in the third delimited
#'   position
#' @return Numeric, latitude in decimal degrees.
#' @note Inputting a numeric to lat will give incorrect results for latitude -10
#'   < x < 10 due to loss of leading zero(s).
#' @keywords internal
#' @examples
#' lat <- em38:::gga_lat('2729.10198', 'S')
#'
gga_lat <- function(lat = NULL, dir = NULL) {
  # this is potentially stupid depending on how the numbers are stored
  # - need more raw datasets for testing
  # its safe if leading zeros are used correctly
  # https://www.labsat.co.uk/index.php/en/free-gps-nmea-simulator-software maybe
  degrees <- as.integer(substr(lat, 1, 2))
  dec_minutes <- as.numeric(substr(lat, 3, nchar(lat)))

  # 7 dec pl is about a centimetre on-ground
  if(dir == 'S') {
    round(0 - (degrees + dec_minutes/60), 7)
  } else {
    round(degrees + dec_minutes/60, 7)
  }
}

#' G*GGA longitude
#'
#' This function retrieves longitude position information from a G*GGA string.
#' @param long String containing longitude information. G*GGA strings are
#'   comma-separated, longitude is in the fourth delimited position
#' @param dir String determining whether longitude is E or W of 0 degrees. GPGGA
#'   strings are comma-separated, EW status is in the fifth delimited position
#' @return Numeric, longitude in decimal degrees.
#' @note Inputting a numeric to long will give incorrect results for longitude
#'   -100 < x < 100 due to loss of leading zero(s).
#' @keywords internal
#' @examples
#' lat <- em38:::gga_long('15257.5556', 'E')
#'
gga_long <- function(long = NULL, dir = NULL) {

  degrees <- as.integer(substr(long, 1, 3))
  dec_minutes <- as.numeric(substr(long, 4, nchar(long)))

  if(dir == 'W') {
    round(0 - (degrees + dec_minutes/60), 7)
  } else {
    round(degrees + dec_minutes/60, 7)
  }
}

#' Convert timestamp to date time object
#'
#' Converts the timestamp on an \*.n38 record to computer time
#' @param comp_time POSIXlt object
#' @param init_stamp state of program timer at comp_time
#' @param now_stamp current state of program timer
#' @return POSIXlt with correct date/time
#' @keywords internal
#'
conv_stamp <- function(comp_time = NULL, init_stamp = NULL, now_stamp) {
  diff_stamp <- now_stamp - init_stamp
  diff_seconds <- diff_stamp / 1000
  comp_time + diff_seconds
}


R/GPS_data.R
#' Calculate NMEA-0183 XOR checksum
#'
#' This function calculates the checksum for NMEA-0183 G*GGA strings.
#' @param string A string with valid NMEA-0183 G*GGA structure.
#' @return Logical; TRUE if checksum is correct.
#' @examples
#' # first G*GGA msg from data('n38_demo')
#' msg_1 <- "$GPGGA,015808.00,2726.53758,S,15126.05255,E,1,08,1.0,365.1,M,39.5,M,,*79"
#' msg_2 <- "$GPG5808.00,2726.53758,S,15126.05255,E,1,08,1.0,365.1,M,39.5,M,,*79"
#' chk_1 <- em38:::nmea_check(string = msg_1)
#' chk_2 <- em38:::nmea_check(string = msg_2)
#'
nmea_check <- function(string = NULL) {
  prs <- unlist(strsplit(string, '\\*'))[1]
  # lower case required for comparison
  chk <- tolower(unlist(strsplit(string, '\\*'))[2])
  # if the checksum signifier * is missing, that's a fail anyway
  if(is.na(chk)) { return(FALSE) }
  # fn still works if starting $ missing already
  prs <- gsub('^\\$', '', prs)
  prs <- sapply(unlist(strsplit(prs, '*')), charToRaw)

  prs_chksum <- base::Reduce(xor, prs)

  # good ol' coercion rules, see ?base::Comparison
  if(prs_chksum == chk) { TRUE } else { FALSE }
}

#' Process NMEA-0183 G*GGA messages
#'
#' This function pulls out position fix data from NMEA-0183 G*GGA strings and
#' dumps it into a list.
#' @param string A string with valid NMEA-0183 G*GGA structure.
#' @return A list containing the data elements recorded in NMEA-0183 G*GGA data
#'   chunks. Elements are given appropriate R data types.
#' @examples
#' # first GPGGA msg from data('n38_demo')
#' msg_1   <- "$GPGGA,015808.00,2726.53758,S,15126.05255,E,1,08,1.0,365.1,M,39.5,M,,*79"
#' gpgga_1 <- em38:::process_gga(string = msg_1)
#' @importFrom units as_units
#'
process_gga <- function(string = NULL) {
  gga_reading <- unlist(strsplit(string, split = c(',')))

  out        <- vector('list', length = 11)
  names(out) <- c('UTC_time', 'latitude', 'longitude', 'fix_quality', 'n_sats',
                  'HDOP', 'antenna_alt', 'geoid_sep', 'difcor_age_s',
                  'base_stn', 'checksum')

  # ignore date, time itself is fine. don't need lt either
  out[['UTC_time']]  <- as.POSIXct(gga_reading[2], format = '%H%M%OS',
                                   tz = 'UTC')

  # see signal_conversion.R for lat/long retrieval. inputs shld be WGS84
  out[['latitude']]  <- gga_lat(gga_reading[3],gga_reading[4])
  out[['longitude']] <- gga_long(gga_reading[5],gga_reading[6])
  out[['fix_quality']]   <- gga_reading[7] # allowed: 0-9
  out[['n_sats']]    <- as.integer(gga_reading[8])
  out[['HDOP']]      <- as.numeric(gga_reading[9])
  out[['antenna_alt']] <- as.numeric(gga_reading[10])
  # antenna_alt is always in metres afaict
  units(out[['antenna_alt']]) <- units::as_units('m')
  out[['geoid_sep']] <- as.numeric(gga_reading[12])
  # same here
  units(out[['geoid_sep']]) <- units::as_units('m')
  # only in use if differential gps is (fix_quality == '2', possibly some other
  # vals as well)
  out[['difcor_age_s']] <-  as.numeric(gga_reading[14])
  # only in use if differential GPS is
  # and it isn't properly comma delimited jfc whyyyyyy >:-(
  out[['base_stn']] <- if(substr(gga_reading[15], 1, 1) == '*') {
    NA
  } else {
    # grab whatever is before the *
    gsub('\\*.*', '', gga_reading[15])
  }
  # checksum
  # "It is the representation of two hexadecimal characters of an XOR of all
  # characters in the sentence between – but not including – the $ and the *
  # character."
  # https://rietman.wordpress.com/2008/09/25/how-to-calculate-the-nmea-checksum/
  out[['checksum']] <- gsub('^[^\\*]*', '', gga_reading[15])
  # validated later
  out
}

#' Process NMEA-0183 G*VTG messages
#'
#' This function pulls out track made good and speed over ground data from
#' NMEA-0183 G*VTG strings and dumps it into a list.
#' @param string A string with valid NMEA-0183 G*VTG structure.
#' @return A list containing 6 data elements recorded in NMEA-0183 G*VTG data
#'   chunks. Elements are given appropriate data types.
#' @examples
#' # first GPVTG msg from data('n38_demo')
#' msg_1   <- "$GPVTG,208.02,T,,M,0.32,N,0.59,K,A*38"
#' gpvtg_1 <- em38:::process_vtg(string = msg_1)
#' @importFrom units as_units
#'
process_vtg <- function(string = NULL) {
  vtg_reading <- unlist(strsplit(string, split = c(',')))

  out        <- vector('list', length = 6)
  names(out) <- c('TMG_Mag', 'TMG_True', 'speed_knots', 'speed_kmh', 'fix_type',
                  'checksum')

  # not sure how to best units the first three
  out[['TMG_Mag']] <- as.numeric(vtg_reading[2])
  out[['TMG_True']] <- as.numeric(vtg_reading[4])
  out[['speed_knots']] <- as.numeric(vtg_reading[6])
  out[['speed_kmh']] <- as.numeric(vtg_reading[8])
  units(out[['speed_kmh']]) <- units::as_units('km/h')
  # http://www.gpsinformation.org/dale/nmea.htm#2.3
  # fix type: A = autonomous, D = differential, E = Estimated, N = not valid,
  # S = Simulator
  out[['fix_type']] <- gsub('\\*.*', '', vtg_reading[10])
  out[['checksum']] <- gsub('^[^\\*]*', '', vtg_reading[10])
  out
}

#' Process NMEA-0183 G*RMC messages
#'
#' This function pulls out reccommended minimum sentence data from NMEA-0183
#' G*RMC strings and dumps it into a list.
#' @param string A string with valid NMEA-0183 G*RMC structure.
#' @return A list containing 9 data elements recorded in NMEA-0183 G*RMC data
#'   chunks. Elements are given appropriate data types.
#' @examples
#' # first GPRMC msg from data('n38_demo')
#' msg_1   <- "$GPRMC,015808.00,A,2726.53758,S,15126.05255,E,0.32,208.02,160318,,,A*48"
#' gprmc_1 <- em38:::process_rmc(string = msg_1)
#'
process_rmc <- function(string = NULL) {
  rmc_reading <- unlist(strsplit(string, split = c(',')))

  out        <- vector('list', length = 10)
  # http://aprs.gids.nl/nmea/#rmc
  names(out) <- c('UTC_date_time', 'validity', 'latitude', 'longitude',
                  'speed_knots', 'TMG_Mag', 'mag_var', 'mag_var_dir',
                  'fix_type', 'checksum')

  out[['UTC_date_time']] <- as.POSIXlt(rmc_reading[2], format = '%H%M%OS',
                                       tz = 'UTC')
  # amend date, WARNING not sure about timezone effects here, not sure if date
  # is supplied in UTC as time is, or if its locale-dependant.
  date <- as.POSIXlt(rmc_reading[10], format = '%d%m%y')
  out[['UTC_date_time']]$mday <- date$mday
  out[['UTC_date_time']]$mon  <- date$mon
  out[['UTC_date_time']]$year <- date$year
  out[['UTC_date_time']]$wday <- date$wday
  out[['UTC_date_time']]$yday <- date$yday
  out[['UTC_date_time']] <- as.POSIXct(out[['UTC_date_time']])

  out[['validity']] <- switch(rmc_reading[3],
                              "A" = "ok",
                              "V" = "invalid")
  # see signal_conversion.R for lat/long retrieval. inputs shld be WGS84
  out[['latitude']]  <- gga_lat(rmc_reading[4],rmc_reading[5])
  out[['longitude']] <- gga_long(rmc_reading[6],rmc_reading[7])
  out[['speed_knots']] <- as.numeric(rmc_reading[8])
  out[['TMG_Mag']] <- as.numeric(rmc_reading[9])
  out[['mag_var']] <- as.numeric(rmc_reading[11])
  out[['mag_var_dir']] <- as.character(rmc_reading[12])
  out[['fix_type']] <- gsub('\\*.*', '', rmc_reading[13])
  out[['checksum']] <- gsub('^[^\\*]*', '', rmc_reading[13])
  out
}

#' Process NMEA-0183 G*GSA messages
#'
#' This function pulls out GPS DOP and active satellites data from NMEA-0183
#' G*GSA strings and dumps it into a list.
#' @param string A string with valid NMEA-0183 G*GSA structure.
#' @return A list containing n data elements recorded in NMEA-0183 G*GSA data
#'   chunks. Elements are given appropriate data types.
#' @examples
#' # first GPGSA msg from data('n38_demo')
#' msg_1   <- "$GPGSA,M,3,05,10,15,16,20,21,26,29,,,,,1.6,1.0,1.2*32"
#' gpgsa_1 <- em38:::process_gsa(string = msg_1)
#'
process_gsa <- function(string = NULL) {
  gsa_reading <- unlist(strsplit(string, split = c(',')))

  out        <- vector('list', length = 8)
  names(out) <- c('fix_selection_mode', 'fix_type', 'sats_used', 'n_sats',
                  'PDOP', 'HDOP', 'VDOP', 'checksum')

  out[['fix_selection_mode']] <- switch(gsa_reading[2],
                                        'M' = 'manual',
                                        'A' = 'automatic')
  out[['fix_type']] <- switch(gsa_reading[3],
                              '1' = 'no fix',
                              '2' = '2D',
                              '3' = '3D')
  # https://www.trimble.com/OEM_ReceiverHelp/V4.44/en/NMEA-0183messages_GSA.html
  # PRN number, 01 through 32 for GPS, 33 through 64 for SBAS, 64+ for GLONASS
  out[['sats_used']] <- as.numeric(gsa_reading[4:15])[!is.na(as.numeric(gsa_reading[4:15]))]
  # saves having to check other message types
  out[['n_sats']] <- length(out[['sats_used']])
  # smaller is better. TODO: rate DOPs
  out[['PDOP']] <- as.numeric(gsa_reading[16])
  out[['HDOP']] <- as.numeric(gsa_reading[17])
  out[['VDOP']] <- as.numeric(gsub('\\*.*', '', gsa_reading[18]))
  out[['checksum']] <- gsub('^[^\\*]*', '', gsa_reading[18])
  out
}


#' Process NMEA-0183 G*GSV messages
#'
#' This function pulls out satellites in view data from NMEA-0183 G*GSV strings
#' and dumps it into a list.
#' @param string A string with valid NMEA-0183 G*GSV structure.
#' @return A list containing 15 data elements recorded in NMEA-0183 G*GSV data
#'   chunks. Elements are given appropriate data types. Note that depending on
#'   the number of satellites in view, up to three of these messages may exist
#'   for every GPS reading. Up to four satellites are reported on per string.
#'   Note also that SNR is receiver-dependant and should only be considered
#'   relative to other readings in the same dataset.
#' @examples
#' # first GPGSV msg from data('n38_demo')
#' msg_1   <- "$GPGSV,3,1,11,05,14,138,46,10,14,316,37,12,04,012,,13,24,100,*76"
#' gpgsv_1 <- em38:::process_gsv(string = msg_1)
#'
process_gsv <- function(string = NULL) {
  gsv_reading <- unlist(strsplit(string, split = c(',')))

  out        <- vector('list', length = 19)
  names(out) <- c('sentence_x_of_y', 'n_sats_in_view',
                  'sat_1_PRN', 'sat_1_elevation', 'sat_1_azimuth', 'sat_1_SNR',
                  'sat_2_PRN', 'sat_2_elevation', 'sat_2_azimuth', 'sat_2_SNR',
                  'sat_3_PRN', 'sat_3_elevation', 'sat_3_azimuth', 'sat_3_SNR',
                  'sat_4_PRN', 'sat_4_elevation', 'sat_4_azimuth', 'sat_4_SNR',
                  'checksum')

  out[['sentence_x_of_y']] <- paste0(gsv_reading[3], ' of ', gsv_reading[2])
  # nb not the same as sats used!
  out[['n_sats_in_view']]  <- as.numeric(gsv_reading[4])
  out[['sat_1_PRN']]       <- if(gsv_reading[ 5] == '') { NA } else { gsv_reading[ 5] }
  out[['sat_2_PRN']]       <- if(gsv_reading[ 9] == '') { NA } else { gsv_reading[ 9] }
  out[['sat_3_PRN']]       <- if(gsv_reading[13] == '') { NA } else { gsv_reading[13] }
  out[['sat_4_PRN']]       <- if(gsv_reading[17] == '') { NA } else { gsv_reading[17] }
  out[['sat_1_elevation']] <- as.numeric(gsv_reading[ 6])
  out[['sat_2_elevation']] <- as.numeric(gsv_reading[10])
  out[['sat_3_elevation']] <- as.numeric(gsv_reading[14])
  out[['sat_4_elevation']] <- as.numeric(gsv_reading[18])
  out[['sat_1_azimuth']]   <- as.numeric(gsv_reading[ 7])
  out[['sat_2_azimuth']]   <- as.numeric(gsv_reading[11])
  out[['sat_3_azimuth']]   <- as.numeric(gsv_reading[15])
  out[['sat_4_azimuth']]   <- as.numeric(gsv_reading[19])
  out[['sat_1_SNR']]       <- as.numeric(gsv_reading[ 8])
  out[['sat_2_SNR']]       <- as.numeric(gsv_reading[12])
  out[['sat_3_SNR']]       <- as.numeric(gsv_reading[16])
  out[['sat_4_SNR']]       <- as.numeric(gsub('\\*.*', '', gsv_reading[20]))
  out[['checksum']]        <- gsub('^[^\\*]*', '', gsv_reading[20])
  out
}
